Modules/xSharePoint/DSCResources/MSFT_xSPWebAppPolicy/MSFT_xSPWebAppPolicy.psm1

```diff
@@ -56,63 +81,260 @@ function Set-TargetResource
+            Import-Module (Join-Path $ScriptRoot "..\..\Modules\xSharePoint.WebAppPolicy\xSPWebAppPolicy.psm1" -Resolve)
+            
+            foreach ($member in $allMembers) {
+                $policy = $wa.Policies | Where-Object { $_.UserName -eq $member.UserName }
+
+                if ($policy -ne $null) {
+                    # User exists. Check permissions
+                    Write-Verbose -Verbose "User $($member.UserName) exists, checking permissions"
+                    if ($member.ActAsSystemAccount -ne $policy.IsSystemUser) { $policy.IsSystemUser = $member.ActAsSystemAccount }
+
+                    $polbinddiff = Compare-Object -ReferenceObject $policy.PolicyRoleBindings.Name -DifferenceObject $member.PermissionLevel
+                    if ($polbinddiff -ne $null) {
+                        $policy.PolicyRoleBindings.RemoveAll()
+                        foreach ($permissionLevel in $member.PermissionLevel) {
+                            switch ($permissionLevel) {
+                                "Deny All" {
+                                    $policy.PolicyRoleBindings.Add($denyAll)
+                                }
+                                "Deny Write" {
+                                    $policy.PolicyRoleBindings.Add($denyWrite)
+                                }
+                                "Full Control" {
+                                    $policy.PolicyRoleBindings.Add($fullControl)
+                                }
+                                "Full Read" {
+                                    $policy.PolicyRoleBindings.Add($fullRead)
+                                }
+                            }
+                        }
+                    }
+                } else {
+                    # User does not exist. Add user
+                    Write-Verbose -Verbose "Adding $($member.UserName)"
+                    $newPolicy = $wa.Policies.Add($member.UserName, $member.UserName)
+                    foreach ($permissionLevel in $member.PermissionLevel) {
+                        switch ($permissionLevel) {
+                            "Deny All" {
+                                $newPolicy.PolicyRoleBindings.Add($denyAll)
+                            }
+                            "Deny Write" {
+                                $newPolicy.PolicyRoleBindings.Add($denyWrite)
+                            }
+                            "Full Control" {
+                                $newPolicy.PolicyRoleBindings.Add($fullControl)
+                            }
+                            "Full Read" {
+                                $newPolicy.PolicyRoleBindings.Add($fullRead)
+                            }
+                        }
+                    }
+                    if ($member.ActAsSystemAccount) {
+                        $newPolicy.IsSystemUser = $member.ActAsSystemAccount
+                    }
+                }
+                $wa.Update()
```

Comment:
Did you see issues calling Update() multiple times on the web app? You should be able to call it just once at the end of adding all the accounts and making changes rather than once at the bottom of the foreach loop there